### ISSUE TITLE
heavy use of RAM memory - Win10

### ISSUE THREAD

**Issue by mrr07** [beekeeper-studio/beekeeper-studio#263]
after running different queries, beekeeper uses a lot of memory, about 2 GB !!, and then crashes

![bee](https://user-images.githubusercontent.com/69007625/88948231-f3c3ae80-d291-11ea-8852-2c2c6bc09051.jpg)

**Comment by rathboma** [comment]
Hey! Thanks for bringing this up, I haven't seen this behavior. Let's fix it :-)

Some questions:

Did you have multiple tabs open with large query results in each, or just running one query tab over and over?

How large were the query results? In rows / columns

**Comment by mrr07** [comment]
I simply write a query in a tab and run it. Usually I don't have more than 5/6 tabs open. Queries are quite large, we are talking about 50000-200000 records but with few columns, maximum 12. For example, after running the following query
![image](https://user-images.githubusercontent.com/69007625/89010465-ea2c5c00-d30e-11ea-8821-2ea61de225c3.png)
I have a memory usage of about one GB
![image](https://user-images.githubusercontent.com/69007625/89010577-28298000-d30f-11ea-8bcc-ada32ee6b334.png)

**Comment by rathboma** [comment]
Yep. I think this is due to the large result size. I have plans to fix this by sending query results to temporary files then paging them in as needed. That would keep the records in memory to ~100

As you clear query results the OS should start reclaiming memory

**Comment by mrr07** [comment]
good idea!!!

**Comment by mdc-git** [comment]
@rathboma 

https://github.com/beekeeper-studio/beekeeper-studio/issues/366#issuecomment-706590469

**Comment by mdc-git** [comment]
@rathboma 

Further digging revealed: even after closing a tab, the memory of the result is somehow retained, which leads to a memory leak. According to my debugging results, a huge array of strings (columndefininitions/rows) stays on the heap for every query result and is never freed. It even happens when there is no Tabulator table involved at all, so I suspect the connection is not released on tab close or something. BTW I'm on Linux if that matters.

**Comment by mdc-git** [comment]
Cool, thank you. A last thing to note is that when the results are assigned to an reactive property (this.results) the result is getting all the setter and getter functions as well, which makes for a major overhead. Might consider Object.freeze before assigning. 

Some links too:
https://stackoverflow.com/questions/59454894/vue-memory-leak-when-rendered-components-are-removed
https://rast.io/vue-observables-and-large-data-sets

**Comment by rathboma** [comment]
I had 10 mins before a (real work) meeting and ran the profiler while running the same query over and over, then closing the tab.

The app seems to do a solid job of reclaiming the memory after unloading the results

![Screen Shot 2020-10-14 at 12 52 05 PM](https://user-images.githubusercontent.com/279769/96026538-3d954d00-0e1c-11eb-84f6-e4b4553b5b14.png)

The real fix here is to not store all results in memory when running a query. Even when there ISNT a memory leak, if you have 10 tabs, each with 20k records that will be quite a lot of memory bloat.

Instead I need to send results to disk then page in as needed.

Good news is I haven't had too  many bug reports of memory related issues, so this isn't super high priority right now, but for sure want to deal with it.

**Comment by mdc-git** [comment]
I fully agree. I'm just not sure how to sort the results by column in a file based approach without loading all rows into memory again.
Either way, it's strange. I'm just getting a growing heap and run out of memory eventually ( just tested it again with the release AppImage ). Might be an edge case. Anyways thanks for your responses.

**Comment by mdc-git** [comment]
Fixed my problem now with some tricks. Sending the results of a big query (select * from bigtable) to disk with a big resultset yielded a 400mb file in my case. For sure you would not wan't to wait for the download everytime you change the query/do a new one, which took about 30secs to a minute. But maybe I'm missing something.

**Comment by UtechtDustin** [comment]
I dont think you miss something, you HAVE to download the resultset every time.
As the query select changes the resultset.
